path_certs.go

```diff
@@ -0,0 +1,618 @@
+	//path := data.Get("path").(string)
+	//b.Logger().Debug("path = " + path)
+
+	serial := data.Get("serial").(string)
+	b.Logger().Debug("serial = " + serial)
+
+	if len(serial) == 0 {
+		return logical.ErrorResponse("The serial number must be provided"), nil
+	}
+
+	if b.System().ReplicationState().HasState(consts.ReplicationPerformanceStandby) {
+		return nil, logical.ErrReadOnly
+	}
+
+	// We store and identify by lowercase colon-separated hex, but other
+	// utilities use dashes and/or uppercase, so normalize
+	serial = strings.Replace(strings.ToLower(serial), "-", ":", -1)
+
+	return revokeCert(ctx, b, req, serial, false)
+}
+
+// Revokes a cert, and tries to be smart about error recovery
+func revokeCert(ctx context.Context, b *keyfactorBackend, req *logical.Request, serial string, fromLease bool) (*logical.Response, error) {
+	// As this backend is self-contained and this function does not hook into
+	// third parties to manage users or resources, if the mount is tainted,
+	// revocation doesn't matter anyways -- the CRL that would be written will
+	// be immediately blown away by the view being cleared. So we can simply
+	// fast path a successful exit.
+	if b.System().Tainted() {
+		return nil, nil
+	}
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	creds := config.Username + ":" + config.Password
+	encCreds := b64.StdEncoding.EncodeToString([]byte(creds))
+
+	b.Logger().Debug("Closing idle connections")
+	http.DefaultClient.CloseIdleConnections()
+
+	kfId, err := req.Storage.Get(ctx, "kfId/"+serial) //retrieve the keyfactor certificate ID, keyed by sn here
+	if err != nil {
+		b.Logger().Error("Unable to retreive Keyfactor certificate ID for cert with serial: "+serial, err)
+		return nil, err
+	}
+
+	var keyfactorId int
+	err = kfId.DecodeJSON(&keyfactorId)
+
+	if err != nil {
+		b.Logger().Error("Unable to parse stored certificate ID for cert with serial: "+serial, err)
+		return nil, err
+	}
+
+	// set up keyfactor api request
+	url := config.KeyfactorUrl + kf_revoke_path
+	payload := fmt.Sprintf(`{
+		"CertificateIds": [
+		  %d
+		],
+		"Reason": 0,
+		"Comment": "%s",
+		"EffectiveDate": "%s"},
+		"CollectionId": 0
+	  }`, keyfactorId, "via HashiCorp Vault", time.Now().UTC().String())
+	//b.Logger().Debug("Sending revocation request.  payload =  " + payload)
+	httpReq, _ := http.NewRequest("POST", url, strings.NewReader(payload))
+
+	httpReq.Header.Add("x-keyfactor-requested-with", "APIClient")
+	httpReq.Header.Add("content-type", "application/json")
+	httpReq.Header.Add("authorization", "Basic "+encCreds)
+
+	res, err := http.DefaultClient.Do(httpReq)
+	if err != nil {
+		b.Logger().Error("Revoke failed: {{err}}", err)
+		return nil, err
+	}
+	if res.StatusCode != 204 {
+		r, _ := ioutil.ReadAll(res.Body)
```

Comment:
I get a deprecation warning when I use `ioutil`. This line could be converted to `io.ReadAll`